src/compute-client/src/controller/replica.rs

```diff
@@ -344,27 +395,87 @@ struct CollectionState<T> {
     ///
     /// If this is `None`, no metrics are collected.
     metrics: Option<ReplicaCollectionMetrics>,
+    /// Tracks whether this collection is hydrated, i.e., it has produced some initial output.
+    hydration_flag: HydrationFlag,
     /// Time at which this collection was installed.
     created_at: Instant,
     /// Original as_of of this collection.
     as_of: Antichain<T>,
 }
 
 impl<T: Timestamp> CollectionState<T> {
-    /// Apply the given frontier update to the collection metrics.
-    fn observe_frontier_update(&self, frontier: &Antichain<T>) {
-        let Some(metrics) = &self.metrics else { return };
-
-        // If the value of `initial_output_duration_seconds` is greater than 0, that means we have
-        // already observed the output before and have nothing else to do.
-        let initial_output_duration = metrics.initial_output_duration_seconds.get();
-        if initial_output_duration > 0. {
-            return;
+    fn hydrated(&self) -> bool {
+        self.hydration_flag.hydrated
+    }
+}
+
+/// A wrapper type that maintains hydration introspection for a given replica and collection, and
+/// ensures that reported introspection data is retracted when the flag is dropped.
+struct HydrationFlag {
+    replica_id: ReplicaId,
+    collection_id: GlobalId,
+    hydrated: bool,
+    introspection_tx: crossbeam_channel::Sender<IntrospectionUpdates>,
+}
+
+impl HydrationFlag {
+    /// Create a new unset `HydrationFlag` and update introspection.
+    fn new(
+        replica_id: ReplicaId,
+        collection_id: GlobalId,
+        introspection_tx: crossbeam_channel::Sender<IntrospectionUpdates>,
+    ) -> Self {
+        let self_ = Self {
+            replica_id,
+            collection_id,
+            hydrated: false,
+            introspection_tx,
+        };
+
+        let insertion = self_.row();
+        self_.send(vec![(insertion, 1)]);
+
+        self_
+    }
+
+    /// Mark the collection as hydrated and update introspection.
+    fn set(&mut self) {
+        if self.hydrated {
+            return; // nothing to do
         }
 
-        if PartialOrder::less_than(&self.as_of, frontier) {
-            let duration = self.created_at.elapsed().as_secs_f64();
-            metrics.initial_output_duration_seconds.set(duration);
+        let retraction = self.row();
+        self.hydrated = true;
+        let insertion = self.row();
+
+        self.send(vec![(retraction, -1), (insertion, 1)]);
+    }
+
+    fn row(&self) -> Row {
+        Row::pack_slice(&[
+            Datum::String(&self.collection_id.to_string()),
+            Datum::String(&self.replica_id.to_string()),
+            Datum::from(self.hydrated),
+        ])
+    }
+
+    fn send(&self, updates: Vec<(Row, Diff)>) {
+        let result = self
+            .introspection_tx
+            .send((IntrospectionType::ComputeHydrationStatus, updates));
+
+        if result.is_err() {
```

Comment:
You could match on the receiver-dropped error here to encode the comment in code!

Reply:
Unfortunately matching here doesn't make things more readable because the returned error is just generically names `SendError`. The code would look like this:

```
        match result {
            Ok(_) => (),
            Err(crossbeam_channel::SendError(_)) => {
                // ...
            }
        }
```

... which would still need the same comment explaining why the error occurred and why it's fine to ignore it.